app/actions/battle.py

```diff
@@ -0,0 +1,15 @@
+def action(rivals: dict, first: str, second: str) -> dict:
+    # calculation of the knight's hp
+    hp_first = rivals[second]["power"] - rivals[first]["protection"]
+    rivals[first]["hp"] -= hp_first
+    hp_second = rivals[first]["power"] - rivals[second]["protection"]
+    rivals[second]["hp"] -= hp_second

```

Comment:
1. use get method instead of parenthesis everywhere to avoid KeyError
rivals.get(second, {}).get("hp")
2.  create separated variables for rivals.get(first) and rivals.get(second) to not dublicate code